File: easy/beststock/best.py
class Solution:
    def __init__(self):
        self._prices: list[int] = [2, 5, 4, 3, 1, 6, 3, 5, 8]

    def best_stock(self) -> int:
        left, right = 0, 1
        maxProfit = 0

        while right < len(self._prices):
            if self._prices[left] < self._prices[right]:
                profit = self._prices[right] - self._prices[left]
                maxProfit = max(maxProfit, profit)
            else:
                left = right

            right += 1

        return maxProfit


def main():
    solution = Solution()

    result = solution.best_stock()

    print("This is best stock we are passing [2, 5, 4, 3, 1, 6, 3, 5, 8]")
    print("We should return 7, which is the max profit of 1 and 8")
    print(result)


if __name__ == "__main__":
    main()

File: easy/beststock/go/best.go
package main

import (
	"fmt"
)

type Solution struct {
	prices []int
}

func NewSolution() *Solution {
	return &Solution{
		prices: []int{2, 5, 4, 3, 1, 6, 3, 5, 8},
	}
}

func (s *Solution) BestProfit() int {
	left, right := 0, 1
	maxProfit := 0

	for right < len(s.prices) {
		if s.prices[left] < s.prices[right] {
			profit := s.prices[right] - s.prices[left]
			
			if profit > maxProfit {
				maxProfit = profit
			}
		} else {
			left = right
		}

		right++
	}

	return maxProfit
}

func main() {

	solution := NewSolution()

	result := solution.BestProfit()

	fmt.Println("This is the best stock we are passing [2, 5, 4, 3, 1, 6, 3, 5, 8]")
	fmt.Println("This should return 7, which is the max profit of 1 and 8")
	fmt.Printf("%v", result)
}
